### ISSUE TITLE
Approach for Cross Validation and Seeding

### ISSUE THREAD

**Issue by sreevishnu-damodaran** [matjesg/deepflash2#14]
Thanks for this wonderful framework.

It would be really helpful if you can help me with an approach for K-fold cross validation.

With the help of classes `RandomTileDataset` and `TileDataset`, I'm able to sample tiles from tiffle files but, how can I validate by performing k-fold cross validation over the entire training data picking some validation tiles in the first round of training and then picking the other tiles in the subsequent round of training without leakage of data between multiple validation folds.

```
train_ds = RandomTileDataset(files, ...)
valid_ds = TileDataset(files, ...)
```

Also, please let me know if `RandomTileDataset` and `TileDataset` picks tiles without leakage between two datasets. That is, are both the splits exclusive of each other? Is there a chance of some of the validation tiles picked by `TileDataset` to be present in the random tiles picked by `RandomTileDataset`

On a related note, please help me with the required seeds that should be set for the entire pipeline with steps such as dataset/data loader preparation, training, inference, etc.

Thanks in advance.

**Comment by matjesg** [comment]
Hi @sreevishnu-damodaran , 

Thanks for bringing up this issue. In your example, there would possibly be data leakage between the `RandomTileDataset` and `TileDataset`. The `RandomTileDataset` samples 1*`sample_mult` random tiles from each image during one epoch. The `TileDataset` performs *standard* tiling (+overlap via the `padding` parameter) on the images and can be limited to `val_length` randomly sampled validation tiles (using a fixed 'val_seed').

At the current stage, the default way of doing cross validation is simply by splitting `files` into train and validation sets (as implemented in the [`EnsemleLearner`](https://github.com/matjesg/deepflash2/blob/8934d934cb1473dd26d78fde324d01d6f65dcac7/deepflash2/learner.py#L291) class using sklearn). 
This approach ensures no data leakage and between images and gives a good estimate on the generalization of the model. However, if only a few images are available for training (as this is the case in the [HuBMAP-20 competition](https://www.kaggle.com/c/hubmap-kidney-segmentation)), the performance of the single models trained on fewer data points may be degraded compared to models that have seen parts (tiles) of all images.

We are planning to add a more detailed documentation of this package soon, until then, you can find the whole training and inference pipeline implemented in the [`EnsembleLearner`](https://github.com/matjesg/deepflash2/blob/master/deepflash2/learner.py#L246) class. 

If you have any suggestions on how to implement a solution for this issue, let me know :)

**Comment by sreevishnu-damodaran** [comment]
Thanks a lot @matjesg 

That helps. I was doing the same file based split for cross-validation.

Sure. I'll let you know if I get any ideas. As of know, I have just started a few experiments with the HuBMAP-20 competition and deepflash2 is a huge help.
